smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java

```diff
@@ -158,6 +168,47 @@ private void handleWriteResult(AsyncResult<?> ar, Message<?> message, ProducerRe
                 kafkaHeaders);
     }
 
+    private void createOutgoingTrace(Message<?> message, String topic, int partition, Headers headers) {
+        if (configuration.getTracingEnabled()) {
+            Optional<TracingMetadata> tracingMetadata = TracingMetadata.fromMessage(message);
+
+            final Span.Builder spanBuilder = TRACER.spanBuilder(topic)
+                    .setSpanKind(Span.Kind.PRODUCER);
+
+            if (tracingMetadata.isPresent()) {
+                // Handle possible parent span
+                final SpanContext parentSpan = tracingMetadata.get().getPreviousSpanContext();
```

Comment:
That's one of the reasons we didn't like the current Kafka instrumentation for OpenTelemetry: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/TracingIterator.java#L46

There's no guarantee that we process a method with `@Incoming` and then it's `@Outgoing` with such timing that there couldn't have been a different `@Incoming` being received before the other `@Outgoing` is processed, so relying on the context that's on the `TRACER` would lead to issues.

At least that's my understanding of things

Reply:
`getPreviousSpanContext()` will return the `SpanContext`, if there is one, from any `@Incoming` that was processed